intro/src/foo/numbers.md

```diff
@@ -0,0 +1,112 @@
+# Parsing Numbers
+
+Chumsky is a 'parser combinator' library. It allows the creation of parsers by combining together many smaller
+parsers. The very smallest parsers are called 'primitives' and live in the
+[`primitive`](https://docs.rs/chumsky/1.0.0-alpha.4/chumsky/primitive/index.html) module.
+
+We're going to want to start by parsing the simplest element of Foo's syntax: numbers.
+
+## Parsing Digit
+<hr>
+
+```rust
+// In `parser`...
+filter(|c: &char| c.is_ascii_digit())
+```
+
+The `filter` primitive allows us to read a single input and accept it if it passes a condition. In our case,
+that condition simply checks that the character is a digit.
+
+If we compile this code now, we'll encounter an error. Why?
+
+Although we promised that our parser would produce an `Expr`, the `filter` primitive only outputs the input
+it found. Right now, all we have is a parser from `char` to `char` instead of a parser from `char` to `Expr`!
+
+To solve this, we need to crack open the 'combinator' part of parser combinators. We'll use Chumsky's `map`
+method to convert the output of the parser to an `Expr`. This method is very similar to its namesake on
+`Iterator`.
+
+```rust
+filter(|c: &char| c.is_ascii_digit())
+    .map(|c| Expr::Num(c.to_digit(10).unwrap() as f64))
+```
+
+Here, we're converting the `char` digit to an `f64` (unwrapping is fine: `map` only gets applied to outputs
+that successfully parsed!) and then wrapping it in `Expr::Num(_)` to convert it to a Foo expression.
+
+Try running the code. You'll see that you can type a digit into `test.foo` and have our interpreter generate
+an AST like so:
+
+```
+Ok(Num(5.0))
+```
+
+## Parsing Digit*s*
+<hr>
+
+If you're more than a little adventurous, you'll quickly notice that typing in a multi-digit number doesn't
+quite behave as expected. Inputting `42` will only produce a `Num(4.0)` AST.
+
+This is because `filter` only accepts a *single* input. But now another question arises: why did our interpreter
+*not* complain at the trailing digits that didn't get parsed?
+
+The answer is that Chumsky's parsers are *lazy*: they will consume all of the input that they can and then stop.
+If there's any trailing input, it'll be ignored.
+
+This is obviously not always desirable. If the user places random nonsense at the end of the file, we want to be
+able to generate an error about it! Worse still, that 'nonsense' could be input the user intended to be part of
+the program, but that contained a syntax error and so was not properly parsed. How can we force the parser to consume
+all of the input?
+
+To do this, we can make use of two new parsers: the `then_ignore` combinator and the `end` primitive.
```

Comment:
Note that this is no longer true: chumsky parsers are now forced to consume everything by default (as if they had an implicit `.then_ignore(end())`). The old behaviour can be recovered with the `.lazy()` combinator. I think this section can be removed from the tutorial entirely.